File: Data Structures & Algorithms/max-water-container/submission-2.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        def calcArea(l, r):
            return min(heights[l], heights[r]) * (r - l)
        
        l, r = 0, len(heights) - 1

        maxArea = 0
        while l < r:
            currLen = calcArea(l, r)
            maxArea = max(maxArea, currLen)

            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1

        return maxArea

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-2.cpp
class Solution {
    unordered_map<char, int>  umap;

    int searchMax(){
        int maxFreq = 0;
        for(const auto& pair: umap){
            maxFreq = max(maxFreq, pair.second);
        }
        return maxFreq;
    }
    
public:
    int characterReplacement(string s, int k) {
        
        // guaranteed minimum
        int maxLen = min(static_cast<int>(s.size()), k);
        int size = 0;
        int start = 0;
        for(int i = 0; i < s.size(); i++){
            umap[s[i]]++;
            size++;

            while (size - searchMax() > k) {
                umap[s[start++]]--;
                size--;
            }
            maxLen = max(maxLen, size);

        }

        return maxLen;
    }
};
